### PR TITLE
[SPARK-36868][SQL] Migrate CreateFunctionStatement to v2 command framework

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
Migrate `CreateFunctionStatement` to v2 command framework

### Why are the changes needed?
Migrate to the standard V2 framework

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing tests

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala [hunk 1]
```diff
@@ -451,7 +451,7 @@ class ResolveSessionCatalog(val catalogManager: CatalogManager)
       val funcIdentifier = identifier.asFunctionIdentifier
       DropFunctionCommand(funcIdentifier.database, funcIdentifier.funcName, ifExists, isTemp)
 
-    case CreateFunctionStatement(nameParts,
+    case CreateFunction(UnresolvedDBObjectName(nameParts, _),
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:454]: I think we should follow `CreateNamespace` and match `ResolvedDBObjectName(catalog, name)` with `if isSessionCatalog(catalog)`

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:454]: let's also add a test that for `CREATE FUNCTION a.b.c`, we should throw a proper error saying that creating function is not supported in v2 catalog.

Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:454]: seems we have a test here https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala#L2213

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala [hunk 2]
```diff
@@ -451,31 +451,34 @@ class ResolveSessionCatalog(val catalogManager: CatalogManager)
...
+        className,
+        resources,
+        true,
+        ignoreIfExists,
+        replace)
+
+    case CreateFunction(ResolvedDBObjectName(catalog, name),
+        className, resources, ignoreIfExists, replace) if isSessionCatalog(catalog) =>
+      CreateFunctionCommand(
+        Some(catalog.name),
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:475]: this is catalog name, not database name

**Comment by huaxingao** [comment]
@cloud-fan Please take one more look? Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 3]
```diff
@@ -4469,13 +4469,23 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
...
+        functionIdentifier,
+        string(ctx.className),
+        resources.toSeq,
+        ctx.EXISTS != null,
+        ctx.REPLACE != null)
+    } else {
+      CreateFunction(
+        UnresolvedDBObjectName(
+          functionIdentifier,
+          isNamespace = true),
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:4483]: why `isNamespace = true`?

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala [hunk 4]
```diff
@@ -483,6 +469,30 @@ class ResolveSessionCatalog(val catalogManager: CatalogManager)
       RefreshFunctionCommand(funcIdentifier.database, funcIdentifier.funcName)
   }
 
+  private def callCreateFunctionCommand(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:472]: nit: `convertToV1CreateFunction`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 3 above)
Comment (huaxingao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:4483]: That's where I got confused. `UnresolvedDBObjectName` is resolved to `ResolvedDBObjectName` only when `isNamespace = true`?
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveCatalogs.scala#L34

Comment (huaxingao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:4483]: Never mind what I said earlier. I will give it one more try.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/LookupCatalog.scala [hunk 5]
```diff
@@ -92,6 +92,31 @@ private[sql] trait LookupCatalog extends Logging {
     }
   }
 
+  object CatalogAndNamespaceForIdentifier {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/LookupCatalog.scala:95]: what's the difference between this and CatalogAndIdentifier ?

**Review by cloud-fan — APPROVED** [review]

**Pull request closed without merging**